[Curl] Use CURLOPT_SSL_EC_CURVES to set key exchange curves

kshukuwa opened the pull request:
#### b262535120dd95bc9d8425b14eb523fa16fe44c5
<pre>
[Curl] Use CURLOPT_SSL_EC_CURVES to set key exchange curves
<a href="https://bugs.webkit.org/show_bug.cgi?id=247933">https://bugs.webkit.org/show_bug.cgi?id=247933</a>

Reviewed by Fujii Hironori.

CURLOPT_SSL_EC_CURVES option to set key exchange curves was
added in 7.73.0. So, we use this option instead of using OpenSSL API.

* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlHandle::enableSSLForHost):
(WebCore::CurlHandle::setSslECCurves):
* Source/WebCore/platform/network/curl/CurlContext.h:
* Source/WebCore/platform/network/curl/CurlSSLHandle.h:
(WebCore::CurlSSLHandle::cipherList const):
(WebCore::CurlSSLHandle::signatureAlgorithmsList const):
(WebCore::CurlSSLHandle::ecCurves const):
(WebCore::CurlSSLHandle::setCipherList):
(WebCore::CurlSSLHandle::setSignatureAlgorithmsList):
(WebCore::CurlSSLHandle::setECCurves):
(WebCore::CurlSSLHandle::getCipherList const): Deleted.
(WebCore::CurlSSLHandle::getSignatureAlgorithmsList const): Deleted.
(WebCore::CurlSSLHandle::getCurvesList const): Deleted.
(WebCore::CurlSSLHandle::setCurvesList): Deleted.
* Source/WebCore/platform/network/curl/CurlSSLVerifier.cpp:
(WebCore::CurlSSLVerifier::CurlSSLVerifier):
* Source/WebCore/platform/network/playstation/CurlSSLHandlePlayStation.cpp:
(WebCore::CurlSSLHandle::platformInitialize):

Canonical link: <a href="https://commits.webkit.org/256692@main">https://commits.webkit.org/256692@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec826dac328ce540a01ebb497b33deba12a2ed45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5800 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29618 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106072 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166414 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5986 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34538 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102788 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4461 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83146 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31429 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-css.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88188 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74334 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40271 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37935 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21079 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4645 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4203 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40352 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/network/curl/CurlContext.cpp:
```diff
@@ -27,7 +27,6 @@
 #include "config.h"
 #include "CurlContext.h"
 
-#if USE(CURL)
```
I think we should keep `#if USE(CURL)`. IIRC, I heard that it's a good practice to keep source code compil-able for all ports.

fujii commented on Source/WebCore/platform/network/curl/CurlContext.cpp:
```diff
@@ -342,10 +341,12 @@ void CurlHandle::enableSSLForHost(const String& host)
 
-    const auto& cipherList = sslHandle.getCipherList();
-    if (!cipherList.isEmpty())
+    if (const auto& cipherList = sslHandle.cipherList(); !cipherList.isEmpty())
```
This is checking `cipherList.isEmpty()`. But, the default value of CURLOPT_SSL_CIPHER_LIST is `NULL`. So, I think it's OK to set emtpy string ("").
I don't think we need a local variable `cipherList` here because it's a simple getter.

kshukuwa replied:
`NULL` means to use OpenSSL's default value. On the other hand, an empty string means that no cipher is set. These values have distinctly different meanings.

Also, the description for CURLOPT_SSL_CIPHER_LIST mentions "one or more cipher strings".
https://curl.se/libcurl/c/CURLOPT_SSL_CIPHER_LIST.html
```
The list must be syntactically correct, it consists of one or more cipher strings separated by colons.
```

kshukuwa replied:
CString::data() returns `nullptr` if it doesn't set any data. So I delete a local variable `cipherList`.

fujii commented on Source/WebCore/platform/network/curl/CurlSSLHandle.h:
```diff
@@ -107,7 +107,7 @@ class CurlSSLHandle {
     String m_cipherList;
     String m_signatureAlgorithmsList;
-    String m_curvesList;
+    String m_ecCurves;
```
Converting UTF-8 to UTF-16 and Converting back to UTF-8 doesn't seems nice. How about using `CString` and UTF-8?
curl handle also retains the CURLOPT_SSL_EC_CURVES string. I don't think `CurlSSLHandle` also should retain the string. But, this doesn't seems like a simple change. Just an idea.

kshukuwa replied:
I use `CString` instead of `String`.

kshukuwa committed:
[Curl] Use CURLOPT_SSL_EC_CURVES to set key exchange curves
https://bugs.webkit.org/show_bug.cgi?id=247933

Reviewed by Fujii Hironori.

CURLOPT_SSL_EC_CURVES option to set key exchange curves was
added in 7.73.0. So, we use this option instead of using OpenSSL API.

* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlHandle::enableSSLForHost):
(WebCore::CurlHandle::setSslECCurves):
* Source/WebCore/platform/network/curl/CurlContext.h:
* Source/WebCore/platform/network/curl/CurlSSLHandle.h:
(WebCore::CurlSSLHandle::cipherList const):
(WebCore::CurlSSLHandle::signatureAlgorithmsList const):
(WebCore::CurlSSLHandle::ecCurves const):
(WebCore::CurlSSLHandle::setCipherList):
(WebCore::CurlSSLHandle::setSignatureAlgorithmsList):
(WebCore::CurlSSLHandle::setECCurves):
(WebCore::CurlSSLHandle::getCipherList const): Deleted.
(WebCore::CurlSSLHandle::getSignatureAlgorithmsList const): Deleted.
(WebCore::CurlSSLHandle::getCurvesList const): Deleted.
(WebCore::CurlSSLHandle::setCurvesList): Deleted.
* Source/WebCore/platform/network/curl/CurlSSLVerifier.cpp:
(WebCore::CurlSSLVerifier::CurlSSLVerifier):
* Source/WebCore/platform/network/playstation/CurlSSLHandlePlayStation.cpp:
(WebCore::CurlSSLHandle::platformInitialize):

Canonical link: https://commits.webkit.org/256692@main
